### PR TITLE
Enable OpenAMP Demo configuration via CMake

### PR DESCRIPTION
**Pull request by bentheredonethat**
Currently the demos' interrupt, shared memory and IPC structure values are hard-coded.

Allow these to configured via CMake configure by 
(a) wrapping C symbols in #ifdef / #ifndef
(b) enabling linker script defined resource table location to be set in CMake.

### CODE REVIEW

File: apps/CMakeLists.txt [hunk 1]
```diff
@@ -8,6 +8,11 @@ set (APPS_ROOT_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
 
 set (APPS_SHARE_DIR "${CMAKE_CURRENT_BINARY_DIR}/share")
 
+# If not provided, add default value for resource table
+if(NOT DEFINED ${RSC_TABLE})
+  set(RSC_TABLE 0x3ed20000 CACHE STRING "")
+endif(NOT DEFINED ${RSC_TABLE})
+
```
Comment (arnopo) [inline apps/CMakeLists.txt:15]: This part is related to the machine, the resource table is optional

**Review by arnopo — CHANGES_REQUESTED** [review]

File: apps/CMakeLists.txt (hunk 1 above)
Comment (bentheredonethat) [reply apps/CMakeLists.txt:15]: ok thanks for the review. is it if we keep this but move it to the machine-specific CMakeLists.txt files?

Comment (bentheredonethat) [reply apps/CMakeLists.txt]: updated so that now this logic is only in the machine

**Review by arnopo — CHANGES_REQUESTED** [review]

**Comment by bentheredonethat** [comment]
@arnopo  updated so that only linker scripts and the 2 CMakeLists.txt files are updated. thanks

File: apps/machine/zynqmp_r5/CMakeLists.txt [hunk 2]
```diff
@@ -3,3 +3,7 @@ collect (APP_COMMON_SOURCES rsc_table.c)
 collect (APP_COMMON_SOURCES zynqmp_r5_a53_rproc.c)
 collect (APP_INC_DIRS "${CMAKE_CURRENT_SOURCE_DIR}")
 
+# If not provided, add default value for resource table
+if(NOT DEFINED ${RSC_TABLE})
+  set(RSC_TABLE 0x3ed20000 CACHE STRING "")
+endif(NOT DEFINED ${RSC_TABLE})
```
[the remark this answers is not in the document]
Comment (bentheredonethat) [reply apps/machine/zynqmp_r5/CMakeLists.txt]: moved to the file you suggestd

File: apps/examples/echo/CMakeLists.txt [hunk 3]
```diff
@@ -38,7 +38,7 @@ foreach (_app rpmsg-echo-ping rpmsg-echo)
       add_executable (${_app}.out ${_sources})
       set_source_files_properties(${_sources} PROPERTIES COMPILE_FLAGS "${_cflags}")
 
-      target_link_libraries(${_app}.out -Wl,-Map=${_app}.map -Wl,--gc-sections ${_linker_opt} -Wl,--start-group ${OPENAMP_LIB}-static ${_deps} -Wl,--end-group)
+      target_link_libraries(${_app}.out -Wl,-Map=${_app}.map  -Wl,--defsym,_rsc_table=${RSC_TABLE} -Wl,--gc-sections ${_linker_opt} -Wl,--start-group ${OPENAMP_LIB}-static ${_deps} -Wl,--end-group)
```
[the remark this answers is not in the document]
Comment (bentheredonethat) [reply apps/examples/echo/CMakeLists.txt]: updated so that it is in this file

File: apps/machine/zynqmp/platform_info.c [hunk 4]
```diff
@@ -36,28 +36,48 @@
...
 #define DEV_BUS_NAME        "platform" /* device bus name. "platform" bus
                                         * is used in Linux kernel for generic
 					* devices */
 /* libmetal devices names used in the examples.
  * They are platform devices, you find them in Linux sysfs
  * sys/bus/platform/devices */
+#ifndef SHM_DEV_NAME
 #define SHM_DEV_NAME        "3ed20000.shm" /* shared device name */
-
+#endif /* SHM_DEV_NAME */
```
Comment (arnopo) [inline apps/machine/zynqmp/platform_info.c:62]: Do you really need to give a name that depend on the address?
Same for POLL_DEV_NAME and IPI_DEV_NAME

**Review by arnopo — APPROVED** [review]
I wonder if the application folder should not be in a separate repo as code is for demo and test and doesn't request same level of maturity.
@edmooring what is your feeling on this?

That's said for time being, LGTM with just a comment/question

File: apps/machine/zynqmp/platform_info.c (hunk 4 above)
Comment (bentheredonethat) [reply apps/machine/zynqmp/platform_info.c:62]: for linux a name is needed as the device is accessed via UIO and device tree node name

**Comment by edmooring** [comment]
> I wonder if the application folder should not be in a separate repo as code is for demo and test and doesn't request same level of maturity. @edmooring what is your feeling on this?
> 
> That's said for time being, LGTM with just a comment/question

I think having the apps in a separate repository will make it harder for newcomers to work with the code. Their lives are hard enough already:)

File: apps/system/generic/machine/zynqmp_r5/linker_large_text.ld [hunk 5]
```diff
@@ -312,10 +312,10 @@ _SDA2_BASE_ = __sdata2_start + ((__sbss2_end - __sdata2_start) / 2 );
    __undef_stack = .;
 } > psu_r5_tcm_ram_0_S_AXI_BASEADDR
 
-.resource_table 0x3ed20000 : {
+.resource_table _rsc_table : {
 	. = ALIGN(4);
 	*(.resource_table)
-} > psu_ddr_S_AXI_BASEADDR
+}
```
Comment (edmooring) [inline apps/system/generic/machine/zynqmp_r5/linker_large_text.ld:318]: Why remove the ' > psu_ddr_S_AXI_BASEADDR'?

Comment (bentheredonethat) [reply apps/system/generic/machine/zynqmp_r5/linker_large_text.ld:318]: its possible that the demo is having IPC structures outside of DDR. For example TCM.

**Review by edmooring — APPROVED** [review]
Looks OK to me. Seems to work.

**Pull request merged**